core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -87,4 +101,73 @@ public static TypeAnalyzer createTestingTypeAnalyzer(PlannerContext plannerConte
                         new TablePropertyManager(),
                         new AnalyzePropertyManager()));
     }
+
+    private static ExpressionAnalyzer createExpressionAnalyzer(Session session,
+                                                               PlannerContext plannerContext,
+                                                               StatementAnalyzerFactory statementAnalyzerFactory,
+                                                               TypeProvider types)
+    {
+        return new ExpressionAnalyzer(plannerContext, new AllowAllAccessControl(), statementAnalyzerFactory, new Analysis(null, ImmutableMap.of(), OTHERS), session, types, WarningCollector.NOOP);
+    }
+
+    private static int computeExpressionsTreeSize(Iterable<? extends Node> expressions)
+    {
+        int size = 0;
+        for (Node expression : expressions) {
+            size += 1;
+            if (!expression.getChildren().isEmpty()) {
+                size += computeExpressionsTreeSize(expression.getChildren());
+            }
+        }
+        return size;
+    }
+
+    private static String getIdFromSession(Session session)
+    {
+        return session.getTransactionId()
+                .map(TransactionId::toString)
+                .orElse(session.getQueryId().getId());
+    }
+
+    private static class CacheIdentityKey
+    {
+        private final String id;
+        private final Iterable<Expression> expressions;
```

Comment:
The  is mutable, so it's not a good candidate for a cache key.
Did you mean to materialize the list with `ImmutableList.copyOf` ?